activerecord/lib/active_record/collection_cache_key.rb

```diff
@@ -2,30 +2,15 @@ module ActiveRecord
   module CollectionCacheKey
 
     def collection_cache_key(collection = all, timestamp_column = :updated_at) # :nodoc:
-      query_signature = Digest::MD5.hexdigest(collection.to_sql)
-      key = "#{collection.model_name.cache_key}/query-#{query_signature}"
+      model_signature = collection.model_name.cache_key
 
       if collection.loaded?
-        size = collection.size
-        timestamp = collection.max_by(&timestamp_column).public_send(timestamp_column)
+        unique_signature = collection.pluck(primary_key, timestamp_column).flatten.join("-".freeze)
       else
-        column_type = type_for_attribute(timestamp_column.to_s)
-        column = "#{connection.quote_table_name(collection.table_name)}.#{connection.quote_column_name(timestamp_column)}"
-
-        query = collection
-          .select("COUNT(*) AS size", "MAX(#{column}) AS timestamp")
-          .unscope(:order)
-        result = connection.select_one(query)
-
-        size = result["size"]
-        timestamp = column_type.deserialize(result["timestamp"])
+        unique_signature = collection.unscope(:order).pluck(primary_key, timestamp_column).flatten.join("-".freeze)
```

Comment:
@hdabrows Well spotted.

You can see the Postgres failure [in this build](https://travis-ci.org/rails/rails/jobs/93511677)

The real culprit is `pluck`, which for an association that goes through a join table, creates a query that tries to pluck attributes that are not both in the `SELECT DISTINCT` and  `ORDER BY` clauses as [required by Postgres](http://www.postgresql.org/docs/9.0/static/sql-select.html#SQL-DISTINCT).

```
ActiveRecord::StatementInvalid: PG::InvalidColumnReference: ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list
LINE 1: ... "developers_projects"."project_id" = $1 ORDER BY developers...
                                                             ^
: SELECT DISTINCT "developers"."id", "developers"."updated_at" FROM "developers" INNER JOIN "developers_projects" ON "developers"."id" = "developers_projects"."developer_id" WHERE "developers_projects"."project_id" = $1 ORDER BY developers.name desc, developers.id desc
```

To be honest, I have no idea how to pluck what I need without always executing the entire query without `pluck`. 

Any ideas?